ratbagd/ratbagd-button.c

```diff
@@ -454,6 +454,90 @@ static int ratbagd_button_get_action_types(sd_bus *bus,
 	return sd_bus_message_close_container(reply);
 }
 
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct ratbag_button_macro *, ratbag_button_macro_unref);
+
+static int ratbagd_button_get_macro(sd_bus *bus,
+				    const char *path,
+				    const char *interface,
+				    const char *property,
+				    sd_bus_message *reply,
+				    void *userdata,
+				    sd_bus_error *error)
+{
+	struct ratbagd_button *button = userdata;
+	_cleanup_(ratbag_button_macro_unrefp) struct ratbag_button_macro *macro = NULL;
+	int r;
+	unsigned int idx;
+
+	r = sd_bus_message_open_container(reply, 'a', "(uu)");
+	if (r < 0)
+		return r;
+
+	macro = ratbag_button_get_macro(button->lib_button);
+	if (!macro)
+		goto out;
+
+	for (idx = 0; idx < ratbag_button_macro_get_num_events(macro); idx++) {
+		enum ratbag_macro_event_type type;
+		int value;
+
+		type = ratbag_button_macro_get_event_type(macro, idx);
+		switch (type) {
+		case RATBAG_MACRO_EVENT_INVALID:
+			abort();
+			break;
+		case RATBAG_MACRO_EVENT_NONE:
+			goto out;
+		case RATBAG_MACRO_EVENT_KEY_PRESSED:
+		case RATBAG_MACRO_EVENT_KEY_RELEASED:
+			value = ratbag_button_macro_get_event_key(macro, idx);
+			break;
+		case RATBAG_MACRO_EVENT_WAIT:
+			value = ratbag_button_macro_get_event_timeout(macro, idx);
+			break;
+		}
+
+		r = sd_bus_message_append(reply, "(uu)", type, value);
+		if (r < 0)
+			return r;
```

Comment:
Looking at https://github.com/systemd/systemd/blob/master/src/run/run.c -> it seems that it is a pretty common way of exiting in case of errors...